src/cpu.rs

```diff
@@ -527,6 +549,181 @@ impl Cpu {
+    }
+
+    fn dcr(&mut self, reg: Operand) {
+        let res = match reg {
+            Operand::A => {
+                self.registers.a = self.registers.a.wrapping_sub(1);
+                self.registers.a
+            }
+            Operand::B => {
+                self.registers.b = self.registers.b.wrapping_sub(1);
+                self.registers.b
+            }
+            Operand::C => {
+                self.registers.c = self.registers.c.wrapping_sub(1);
+                self.registers.c
+            }
+            Operand::D => {
+                self.registers.d = self.registers.d.wrapping_sub(1);
+                self.registers.d
+            }
+            Operand::E => {
+                self.registers.e = self.registers.e.wrapping_sub(1);
+                self.registers.e
+            }
+            Operand::H => {
+                self.registers.h = self.registers.h.wrapping_sub(1);
+                self.registers.h
+            }
+            Operand::L => {
+                self.registers.l = self.registers.l.wrapping_sub(1);
+                self.registers.l
+            }
+            Operand::M => {
+                let location = self.registers.get_hl() as usize;
+                self.memory[location] = self.memory[location].wrapping_sub(1);
+                self.memory[location]
+            }
+            _ => panic!("INR only accepts registers or a memory location"),
+        };
+        // update flags
+        self.condition_codes.reset_carry();
+        self.condition_codes.set_zero(res);
+        self.condition_codes.set_sign(res);
+        self.condition_codes.set_parity(res);
+        self.condition_codes
+            .set_aux_carry((res.wrapping_add(1) & 0x0F) == 0x00);
+    }
+
+    fn adc(&mut self, val: u8) {
+        let reg_a = self.registers.a;
+        let carry: u8 = if self.condition_codes.carry { 1 } else { 0 };
+        let res = (reg_a as u16)
+            .wrapping_add(val as u16)
+            .wrapping_add(carry as u16);
+        // put result in accumulator
+        self.registers.a = res as u8;
+        // update flags
+        self.condition_codes.set_zero(res as u8);
+        self.condition_codes.set_sign(res as u8);
+        self.condition_codes.set_parity(res as u8);
+        self.condition_codes.set_carry(res > 0xFF);
+        self.condition_codes
+            .set_aux_carry((reg_a & 0x0F) + (val & 0x0F) + (carry & 0x0F) > 0x0F);
+    }
+
+    fn sub(&mut self, val: u8) {
+        let reg_a = self.registers.a;
+        let res: u16 = (reg_a as u16).wrapping_sub(val as u16);
+        // put result in accumulator
+        self.registers.a = res as u8;
+        // update flags
+        self.condition_codes.set_zero(res as u8);
+        self.condition_codes.set_sign(res as u8);
+        self.condition_codes.set_parity(res as u8);
+        self.condition_codes.set_carry(reg_a < val);
+        self.condition_codes
+            .set_aux_carry((reg_a as i8 & 0x0F) - (val as i8 & 0x0F) >= 0);
```

Comment:
I tried thinking of a way to do this without casting to i8, but came up with a blank. I'm sure there are alternatives, but probably not worth spending time on now.

Reply:
That was the only way I could think to make it work, but I didn't try to find a different one